python-lib/wordcloud_visualizer.py

```diff
@@ -0,0 +1,232 @@
+            languages, subcharts = zip(*group_names)
+            languages = list(languages)
+            self.subcharts = list(subcharts)
+        elif self.subchart_column:
+            self.subcharts = group_names
+            languages = [self.language] * len(self.subcharts)
+        else:
+            languages = group_names
+
+        # Tokenize
+        docs = [self.tokenizer.tokenize_list(text, language)[0] for text, language in zip(texts, languages)]
+
+        return docs
+
+    @time_logging(log_message="Counting tokens")
+    def _count_tokens(self, docs: list):
+        """Count tokens in each group"""
+        counters = []
+        for doc in docs:
+            counter = Counter()
+            for token in doc:
+                counter[(token.text)] += 1  # Equivalently, token.lemma_
+            counters.append(counter)
+
+        if not self.subchart_column:
+            # sum the values with same keys
+            counts = Counter()
+            for d in counters:
+                counts.update(d)
+
+            counts = dict(counts)
+            return counts
+        else:
+            counts_df = pd.DataFrame(
+                list(zip(self.subcharts, counters)),
+                columns=["subchart", "count"],
+            )
+            counts_df = counts_df.groupby(by=["subchart"]).agg({"count": "sum"})
+            # remove subcharts emptied by filter
+            counts_df = counts_df.loc[counts_df["count"] != {}, :]
+            return counts_df
+
+    @time_logging(log_message="Generating wordclouds")
+    def generate_wordclouds(self, counts):
+        """Generate wordclouds and save them as png images"""
+        if self.subchart_column:
+            for name, row in counts.iterrows():
+                # Generate file name and chart title
+                output_file_name = pathvalidate.sanitize_filename(
+                    f"wordcloud_{self.subchart_column}_{name}.png"
+                ).lower()
+                wordcloud_title = output_file_name[:-4]
+                # Generate chart
+                if self.language_as_subchart:
+                    fig = self._generate_wordcloud(row["count"], wordcloud_title, name)
+                else:
+                    fig = self._generate_wordcloud(row["count"], wordcloud_title, self.language)
+                # Return chart
+                temp = BytesIO()
+                fig.savefig(temp, bbox_inches=self.bbox_inches, pad_inches=self.pad_inches, dpi=fig.dpi)
+                yield (temp, output_file_name)
+                plt.close()
+        else:
+            # Generate chart
+            fig = self._generate_wordcloud(counts, "wordcloud", self.language)
+            # Return chart
+            temp = BytesIO()
+            fig.savefig(temp, bbox_inches=self.bbox_inches, pad_inches=self.pad_inches, dpi=fig.dpi)
+            yield (temp, "wordcloud.png")
+            plt.close()
+
+    def prepare_and_count(self, df: pd.DataFrame):
```

Comment:
Similar comment: This method is public and important, could you add type hints + docstring?
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Reply:
Could we rename this to `tokenize_and_count` to clarify what this method does? Also, a type hint on the output would be great: `def my_function(self, param: int) -> List[Tuple]`

Reply:
Perhaps add a bit more type hint info in `Tuple`. If I understand correctly, it's actually `List[Tuple[AnyStr, Dict]]`